// pokemontcgsdk/lib/queryBuilder.js
"use strict";

exports.__esModule = true;
exports["default"] = void 0;

var _axios = _interopRequireDefault(require("axios"));

var _qs = _interopRequireDefault(require("qs"));

var _configure = _interopRequireDefault(require("./configure"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

var getOptions = function getOptions() {
  var options = {
    headers: {}
  };
  if (_configure["default"].apiKey) options.headers['X-Api-Key'] = _configure["default"].apiKey;
  return options;
};

var get = function get(type, args) {
  return _axios["default"].get(_configure["default"].host + "/" + type + (args && '?' + _qs["default"].stringify(args)), getOptions()).then(function (response) {
    return response.data;
  });
};

var _default = function _default(type) {
  return {
    find: function find(id) {
      return (0, _axios["default"])(_configure["default"].host + "/" + type + "/" + id, getOptions()).then(function (response) {
        return response.data.data;
      });
    },
    where: function where(args) {
      return get(type, args);
    },
    all: function all(args, data) {
      if (args === void 0) {
        args = {};
      }

      if (data === void 0) {
        data = [];
      }

      var getAll = function getAll(type, args) {
        var page = args.page ? args.page + 1 : 1;
        return get(type, _extends({}, args, {
          page: page
        })).then(function (response) {
          var _data;

          (_data = data).push.apply(_data, response.data);

          if (!response.totalCount || response.pageSize * response.page >= response.totalCount) {
            return data;
          }

          return getAll(type, _extends({}, args, {
            page: page
          }));
        })["catch"](function (error) {
          return console.error(error);
        });
      };

      return getAll(type, args);
    }
  };
};

exports["default"] = _default;
module.exports = exports.default;
